# Task 2_ a model to predict sale

In [20]:
import pandas as pd
# Assuming your dataset is in a CSV file named 'marketing_data.csv'


In [3]:
file_path = r'Y:\A doc. reserve\A_last_sem_1\IS 352 Emerging Information Technologies and Innovations\Machine Learning Tasks\0x01-linear_regression\dataset\04-advertising_dataset.csv'
df = pd.read_csv(file_path)

# Checking for missing values

 print(df.isnull().sum())

# Displaying data types and statistical summary

In [8]:
print(df.info())
print(df.describe())

correlation_matrix = df.corr()
print(correlation_matrix['Sales'].sort_values(ascending=False))

from sklearn.model_selection import train_test_split


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   TV         200 non-null    float64
 1   Radio      200 non-null    float64
 2   Newspaper  200 non-null    float64
 3   Sales      200 non-null    float64
dtypes: float64(4)
memory usage: 6.4 KB
None
               TV       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   30.554000   15.130500
std     85.854236   14.846809   21.778621    5.283892
min      0.700000    0.000000    0.300000    1.600000
25%     74.375000    9.975000   12.750000   11.000000
50%    149.750000   22.900000   25.750000   16.000000
75%    218.825000   36.525000   45.100000   19.050000
max    296.400000   49.600000  114.000000   27.000000
Sales        1.000000
TV           0.901208
Radio        0.349631
Newspaper    0.157960
Name: Sales, dtype: float64


# Use the feature with the highest correlation


In [12]:
X = df[['TV']]  # Use the feature with the highest correlation
y = df['Sales']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Create and train the model

In [13]:
simple_lr_model = LinearRegression()
simple_lr_model.fit(X_train, y_train)

LinearRegression()

# Make predictions

In [15]:
y_pred_simple = simple_lr_model.predict(X_test)

# Evaluate the model

In [18]:
mse_simple = mean_squared_error(y_test, y_pred_simple)
r2_simple = r2_score(y_test, y_pred_simple)

print("Simple Linear Regression - Mean Squared Error:", mse_simple)
print("Simple Linear Regression - R2 Score (Coefficient of Determination):", r2_simple)


Simple Linear Regression - Mean Squared Error: 6.101072906773964
Simple Linear Regression - R2 Score (Coefficient of Determination): 0.802561303423698


# prepare the data for multiple Linear Regression

In [22]:
X_multi = df[['TV', 'Radio', 'Newspaper']]


# split the data into training and testing sets

In [24]:
X_train_multi, X_test_multi, y_train_multi, y_test_multi = train_test_split(X_multi, y, test_size=0.2, random_state=42)


# build  and train the mutiple Linear regression model

In [26]:
multi_lr_model = LinearRegression()
multi_lr_model.fit(X_train_multi, y_train_multi)

LinearRegression()

# Make predictions

In [28]:
y_pred_multi = multi_lr_model.predict(X_test_multi)

# Evaluate the model

In [30]:
mse_multi = mean_squared_error(y_test_multi, y_pred_multi)
r2_multi = r2_score(y_test_multi, y_pred_multi)

print("Multiple Linear Regression - Mean Squared Error:", mse_multi)
print("Multiple Linear Regression - R2 Score (Coefficient of Determination):", r2_multi)


Multiple Linear Regression - Mean Squared Error: 2.907756910271091
Multiple Linear Regression - R2 Score (Coefficient of Determination): 0.9059011844150826


'''
Multiple Linear Regression (MLR) outperforms Simple Linear Regression (SLR) on both metrics:

MSE: MLR has a lower MSE (2.9077) compared to SLR (6.1011), indicating smaller prediction errors.
R-squared: MLR has a higher R-squared (0.9059) compared to SLR (0.8026), suggesting it explains 
more of the variance in the target variable.
Implications:

Better Fit: MLR appears to provide a better overall fit to the data, capturing more complex relationships 
between features and the target variable.
Improved Prediction Accuracy: MLR is likely to yield more accurate predictions on new data compared to SLR.
Multiple Features Matter: The results suggest that considering multiple features, rather than just one, 
significantly improves prediction performance in this case'''